python/lsst/pipe/tasks/dcrModel.py

```diff
@@ -0,0 +1,539 @@
+            Configuration settings for warping an image
+        exposure : `lsst.afw.image.Exposure`, optional
+            The input exposure to build a matched template for.
+            May be omitted if all of the metadata is supplied separately
+        visitInfo : `lsst.afw.image.VisitInfo`, optional
+            Metadata for the exposure. Ignored if ``exposure`` is set.
+        bbox : `lsst.afw.geom.Box2I`, optional
+            Sub-region of the coadd. Ignored if ``exposure`` is set.
+        wcs : `lsst.afw.geom.SkyWcs`, optional
+            Coordinate system definition (wcs) for the exposure.
+            Ignored if ``exposure`` is set.
+        mask : `lsst.afw.image.Mask`, optional
+            reference mask to use for the template image.
+            Ignored if ``exposure`` is set.
+
+        Returns
+        -------
+        templateImage : `lsst.afw.image.maskedImageF`
+            The DCR-matched template
+
+        Raises
+        ------
+        ValueError
+            If ``filterInfo`` is not set.
+        ValueError
+            If neither ``exposure`` or all of ``visitInfo``, ``bbox``, and ``wcs`` are set.
+        """
+        if self.filterInfo is None:
+            raise ValueError("'filterInfo' must be set for the DcrModel in order to calculate DCR.")
+        if exposure is not None:
+            visitInfo = exposure.getInfo().getVisitInfo()
+            bbox = exposure.getBBox()
+            wcs = exposure.getInfo().getWcs()
+        elif visitInfo is None or bbox is None or wcs is None:
+            raise ValueError("Either exposure or visitInfo, bbox, and wcs must be set.")
+        dcrShift = calculateDcr(visitInfo, wcs, self.filterInfo, len(self))
+        templateImage = afwImage.MaskedImageF(bbox)
+        for subfilter, dcr in enumerate(dcrShift):
+            templateImage += applyDcr(self[subfilter][bbox], dcr, warpCtrl)
+        if mask is not None:
+            templateImage.setMask(mask[bbox])
+        return templateImage
+
+    def conditionDcrModel(self, subfilter, newModel, bbox, gain=1.):
+        """Average two iterations' solutions to reduce oscillations.
+
+        Parameters
+        ----------
+        subfilter : `int`
+            Index of the current subfilter within the full band.
+        newModel : `lsst.afw.image.MaskedImage`
+            The new DCR model for one subfilter from the current iteration.
+            Values in ``newModel`` that are extreme compared with the last
+            iteration are modified in place.
+        bbox : `lsst.afw.geom.box.Box2I`
+            Sub-region of the coadd
+        gain : `float`, optional
+            Additional weight to apply to the model from the current iteration.
+            Defaults to 1.0, which gives equal weight to both solutions.
+        """
+        # The models are MaskedImages, which only support in-place operations.
+        newModel *= gain
+        newModel += self[subfilter][bbox]
+        newModel.image.array /= 1. + gain
```

Comment:
Why not just `newModel /= 1.0 + gain`? That will apply it to the image and variance.

Reply:
Ah, yes, it divides the variance plane by `(1.0 + gain)^2`.